raster2dggs/indexers/dggalrasterindexer.py

```diff
@@ -0,0 +1,435 @@
+
+    def parent_cells(self, cells: set, resolution) -> map:
+        """
+        Implementation of interface function.
+        """
+        # TODO appropriately handle potential for multiple parentage in dggal (e.g. ISEAH3)
+        child_resolution = self.dggrs.getZoneLevel(
+            self.dggrs.getZoneFromTextID(next(iter(cells)))
+        )
+        return map(
+            lambda zone: self.dggrs.getZoneTextID(
+                apply_n_reduce(
+                    self.dggrs.getZoneParents,
+                    child_resolution - resolution,
+                    self.dggrs.getZoneFromTextID(zone),
+                )
+            ),
+            cells,
+        )
+
+    def expected_count(self, parent: str, resolution: int):
+        """
+        Implementation of interface function.
+        """
+        return self.cell_to_children_size(parent, resolution)
+
+    def cell_to_point(self, cell: str) -> shapely.geometry.Point:
+        geo_point: dggal.GeoPoint = self.dggrs.getZoneWGS84Centroid(
+            self.dggrs.getZoneFromTextID(cell)
+        )
+        return shapely.Point(geo_point.lon, geo_point.lat)
+
+    def cell_to_polygon(self, cell: str) -> shapely.geometry.Polygon:
+        geo_points: List[dggal.GeoPoint] = self.dggrs.getZoneWGS84Vertices(
+            self.dggrs.getZoneFromTextID(cell)
+        )
+        return shapely.Polygon(tuple([(p.lon, p.lat) for p in geo_points]))
+
+
+# def ISEA7HRasterIndexer(DGGALRasterIndexer):
+#     def __init__(self, dggs: str):
+#         super().__init__(dggs)
+#         self.dggrs = dggal.ISEA7H()
+
+# NB  All zones of GNOSIS Global Grid and ISEA9R have single parents, whereas ISEA3H zones have one parent if they are a centroid child, and three parents otherwise if they are a vertex child.  See dggrs.getMaxParents()
+
+
+class ISEA4RRasterIndexer(DGGALRasterIndexer):
+    """
+    A raster indexer for the ISEA4R DGGS, an equal area rhombic grid with a refinement ratio of 4 defined in the ISEA projection transformed into a 5x6 Cartesian space resulting in axis-aligned square zones.
+    """
+
+    def __init__(self, dggs: str):
+        super().__init__(dggs)
+        dggal.pydggal_setup(dggal.Application(appGlobals=globals()))
+        self._dggrs = dggal.ISEA4R()
+        self._refinementRatio = self.dggrs.getRefinementRatio()
+
+    @property
+    def dggrs(self) -> dggal.DGGRS:
+        return self._dggrs
+
+    @property
+    def refinementRatio(self) -> int:
+        return self._refinementRatio
+
+
+class IVEA4RRasterIndexer(DGGALRasterIndexer):
+    """
+    A raster indexer for the IVEA4R DGGS, an equal area rhombic grid with a refinement ratio of 4 defined in the IVEA projection transformed into a 5x6 Cartesian space resulting in axis-aligned square zones, using the same global indexing and sub-zone ordering as for ISEA4R.
+    """
+
+    def __init__(self, dggs: str):
+        super().__init__(dggs)
+        dggal.pydggal_setup(dggal.Application(appGlobals=globals()))
```

Comment:
Ideally, you should be able to have multiple `DGGALRasterIndexer` with a single `Application()` instantiated and `pydggal_setup()` only called once (e.g., on the first DGGAL indexer setup).

I was not sure whether it was even safe to instantiate multiple `dggal.Application()` classes, but I just did a small test and both DGGRSs still seem to work fine though.

Reply:
In this case there can only be instantiation with each invocation of the command, so there should be no multiple instantiations possible.